Improvements to building "cargo" packages

AndrewGaspar opened the pull request:
This change introduces several features to improve the installation of packages written in Rust.

Fixes #16045 

cc @adamjstewart @hartzell @ax3l 

# Automatic Vendoring of Cargo Dependencies
This change introduces the ability automatically vendor cargo dependencies by way of the new `cargo_manifest` directive.

If the cargo package has locked its dependencies via a "Cargo.lock" file, that is considered to sufficiently check the dependencies for purposes of spack's checksum validation. If there is no "Cargo.lock" file, the user will be prompted to disable checksums, and the `--no-checksum` flag is respected

I still need to implement the caching of the vendored dependencies to support mirroring. See TODO.

## TODO
- [x] Add support for caching of cargo dependencies
- [x] `.cargo/config` must include non-crates.io sources
  - `.cargo/config` file should be derived from the output of `cargo vendor`.
  - Particularly necessary for packages from non-crates.io sources, since they may include dependencies via version control
- [x] Fix bug with cargo package as being listed as "already staged"
  - This happens because the `CargoStage` expands the root stage during its own fetch
  - EDIT: Just output a message indicating why the package is being staged early.

## Limitations
Currently only supports vendoring from a single cargo manifest per-package. Perhaps this should be expanded, but I think it's pretty rare for a single package to vendor Cargo dependencies from multiple, disjoint manifests.

# `CargoPackage`
This PR also includes a new `CargoPackage` base class, which stream-lines building cargo packages. This automatically implements `build` and `install`.

This is NOT implemented using `cargo install` because it only installs bin targets. However, I suspect a common case will be to also install dynamic and static libraries with a C interface. Therefore, I manually implement the install for all "bin", "staticlib", and "cdylib" targets.

Additionally, I provide a new `crates_io` fetch strategy. This makes it possible to directly import packages from "crates.io", which is the first-party package repository for Rust. The goal is to streamline import of Rust packages that people want to use.

## TODO
- [x] Make it easy to specify cargo features via spack variants.

## Limitations
In the future, it may be necessary to provide a way to specify the targets which will be installed.

# Miscellania
Implemented `cargo checksum` for `crates_io` packages.

Fixed a bug in `_determine_extension` to respect `fetcher.extension`.

Add `ytop` package.

Update `bat`, `fd-find`, `ripgrep`, and `rust-bindgen` packages to use `CargoPackage`.

## TODO
- [x] Documentation: https://spack.readthedocs.io/en/latest/build_systems.html
- [x] Testing

Commit:
Add capability to stage vendored cargo dependencies

Commit:
Install default targets in CargoPackage

Commit:
Add a crates_io URL attr to directly import Rust packages from crates.io

Commit:
Add automatic checksumming for crates_io packages

Remove versions of ripgrep which do not have a Cargo.lock file

Commit:
Update fd-find to use CargoPackage

Commit:
bat should use CargoPackage and depend on oniguruma and libgit2

Commit:
Tell rustc to build with rpath

Commit:
Do not checksum yanked packages

Commit:
Make it possible to built unlocked cargo packages (requires confirmation)

Commit:
Update rust-bindgen package to use CargoPackage

Commit:
Add a template for CargoPackage

Commit:
Add back a bunch of ripgrep versions

Commit:
Add ytop package

Commit:
Add a TODO

adamjstewart commented on lib/spack/spack/build_environment.py:
```diff
@@ -438,6 +439,15 @@ def _set_variables_for_single_module(pkg, module):
+    m.cargo_build = copy.deepcopy(m.cargo)
+    m.cargo_build.add_default_arg('build')
+    m.cargo_build.add_default_arg('--jobs')
+    m.cargo_build.add_default_arg(str(jobs))
```
A lot of the stuff in this function is leftover from before Spack had base classes. I would put this stuff in `CargoPackage`, or in the `cargo` package class. We handle `python` and `bazel` similarly I think.

AndrewGaspar replied:
Hm... so if I override `setup_dependent_package`, it results in me changing the final target package's module, not the module of `CargoPackage`, which is where I ultimately want to call `cargo`. Not sure what the correct solution is there, but I think I'll just define something like this in `CargoPackage:
```python
@property
def cargo(self):
    return Executable(join_path(self.spec['rust'].prefix.bin, 'cargo'))
```

adamjstewart commented:
Can you add this to your TODO list: add documentation on the Cargo build system here: https://spack.readthedocs.io/en/latest/build_systems.html

AndrewGaspar commented on lib/spack/spack/stage.py:
```diff
@@ -735,6 +738,115 @@ def cache_local(self):
+
+    def fetch(self, *args, **kwargs):
+        # Fetching the cargo stage requires the package is expanded
+        self.package_stage.expand_archive()
```
Let me know how you feel about this.

Commit:
Add tokei package

Commit:
Fix checksum sorting of versions

Commit:
Add starship package

Commit:
Move cargo out of build_environment

Commit:
Add jql package

Commit:
Add lsd package

Commit:
bat depends on llvm by way on onig-sys (libclang)

Commit:
Make it easy to template new packages from crates.io

Commit:
Add cargo-mpirun package

Commit:
Add du-dust package

Commit:
Add procs package

Commit:
Add sd package

Commit:
Add tealdeer package

Commit:
Get local caching working

AndrewGaspar commented:
I don't really like how I implemented local caching. I eschewed the fetcher interface, but I think that's going to have some downsides. I should find a way to shoehorn "cargo vendor" into the "fetcher" interface - perhaps it should always archive the dependencies so that the "vendor" path and the "cache"/"mirror" path all look the same.

Commit:
Add master branch for rust-bindgen

Commit:
Let crates_io package names differ from their crate name

Commit:
Add git-delta package

Commit:
Only require llvm in bat for the newest version

Commit:
Refine LLVM dependency in rust-bindgen

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

AndrewGaspar committed:
Fix `spack checksum` for crates.io packages

AndrewGaspar committed:
Add new starship versions

AndrewGaspar committed:
Update exa package to use CargoPackage

AndrewGaspar committed:
Add hyperfine package

AndrewGaspar committed:
Formatting for fd-find

AndrewGaspar committed:
Cleanup in starship

AndrewGaspar committed:
flake8

Commit:
Implement cargo package mirroring

Moves the implementation of dependency vendoring into a fetch strategy to
make it easy to fetch dependencies from alt sources, like a mirror.

Commit:
remove unneeded recursive fetch_strategy import

Commit:
Merge remote-tracking branch 'upstream/develop' into cargo-package

Commit:
Fix archive_file and add def path to CargoStage

Commit:
Remove unnecessary which('cargo')

Commit:
Move which('cargo') out of the CargoVendorFetchStrategy init and into the fetch method

Commit:
Fix spack checksum

Commit:
New ytop versions delivered from crates.io

Commit:
New ripgrep version

Commit:
New jql versions

Commit:
Fix "All fetchers failed" message for cargo dependencies

Commit:
Fix --locked application for disabling checksums

Commit:
flake8

Commit:
CargoPackage exposes prefer_dynamic and lto as variants.

prefer_dynamic is the default, and lto is incompatible with
prefer_dynamic. lto is off by default, but supports thin and fat lto.

This changes how packages get built a bit. Now, `cargo rustc`
is used, rather than `cargo build` wiht RUSTFLAGS. This allows
us to only specify the "prefer-dynamic" flag for the final target,
but not for its dependencies. `cargo rustc` can only build a
single target at a time, so now we look through all the workspace
targets and build them one at a time (dependencies are shared).

Updates some packages to change the defaults for prefer_dynamic
and lto for packages that don't build with prefer-dynamic or prefer
lto.

Also fixes what looked like it would be a bug for staticlib/cdylib
install support.

Commit:
New package versions

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

Commit:
Derive "CargoVendorFetchStrategy.cachable" from package fetcher

Commit:
Address compatibility issues with python 2

AndrewGaspar committed:
Remove llvm dependency for bat in new versions

AndrewGaspar committed:
Add new oniguruma version

hartzell reviewed (commented):
It looks like you've added every version of the various cargo-using applications that you could find.

I thought that the Spack policy was to add the things that you actually needed.

AndrewGaspar commented:
Oh, we can trim it down, if so - I just copied the output of `cargo checksum`.

hartzell commented:
See if any of the core folks chime in; I'm just echoing what I've heard @adamjstewart and others say in other work.

AndrewGaspar committed:
Factor out compiler_entry selection into a separate routine from optimization_flags

AndrewGaspar committed:
Change "power*" names to "pwr*". Clang aliases power* to pwr* and Rust only presents pwr*

AndrewGaspar committed:
Move "rust_archs" to a separate module to share with cargo

AndrewGaspar committed:
Apply microarchitecture flags to CargoPackage builds

This pulls the name of the cpu from the name used for clang
and applies it to rust using "-C target-cpu=foo". The target triple is
also specified now because that has the side effect of not applying
RUSTFLAGS to build scripts, which is the desired behavior for
cross-compiling

AndrewGaspar committed:
Add support for -j to rust package

AndrewGaspar committed:
Add missing pkgconfig dependencies

Commit:
Optimize Rust package for target microarchitecture

AndrewGaspar committed:
libgit2 v1.0.1

Commit:
Rust 1.44.0

Commit:
Merge branch 'spack-rust-1.44.0' into cargo-package

AndrewGaspar committed:
Add starship 0.42.0

AndrewGaspar committed:
Merge remote-tracking branch 'upstream/develop' into cargo-package

adamjstewart commented:
> See if any of the core folks chime in; I'm just echoing what I've heard @adamjstewart and others say in other work.

Yes, I prefer to only add the versions you need, at least until #15896 is resolved. Currently we don't have a good way of removing unused versions of a package, so a change to a build system/patch can be really painful. Of course, if you're willing to maintain these packages and all of these versions, I'm more than happy to let you add every version. I did the same for things like Bazel and TF only because I'm willing to maintain them (for now 😉).

AndrewGaspar commented:
OK - I'll remove all but the newest versions before the merge.

AndrewGaspar committed:
Add support for selecting cargo features

AndrewGaspar committed:
Fix test failure due to change from power9 march to more universal pwr9

AndrewGaspar committed:
flake8

AndrewGaspar committed:
Merge remote-tracking branch 'upstream/develop' into cargo-package

AndrewGaspar committed:
Add newest tokei version

AndrewGaspar committed:
Add documentation for CargoPackage

AndrewGaspar committed:
Remove old versions of bat

AndrewGaspar committed:
Remove old versions of cargo-mpirun and add mpi as a run dependency

AndrewGaspar committed:
Remove old versions of delta

AndrewGaspar committed:
Remove old versions of du-dust

AndrewGaspar committed:
Remove old version of exa

AndrewGaspar committed:
Remove old versions of fd-find

AndrewGaspar committed:
Remove old versions of hyperfine

AndrewGaspar committed:
Remove old versions of jql

AndrewGaspar committed:
Remove old versions of lsd

AndrewGaspar committed:
Remove old versions of procs

AndrewGaspar committed:
ripgrep v12.1.1 - remove old versions

AndrewGaspar committed:
Add AndrewGaspar as maintainer to ripgrep

AndrewGaspar committed:
Remove old versions of rust-bindgen

AndrewGaspar committed:
Remove old versions of sd

AndrewGaspar committed:
starship v0.43.0 - remove old versions

AndrewGaspar committed:
Remove old versions of tealdeer

AndrewGaspar committed:
tokei 12.0.4

AndrewGaspar committed:
Remove old versions of ytop

AndrewGaspar committed:
Add AndrewGaspar as a maintainer of fd-find, jql, and lsd

AndrewGaspar committed:
Add pcre2 variant to ripgrep

AndrewGaspar committed:
Enable pcre2 variant by default

AndrewGaspar committed:
Add nightly-only simd_accel variant to ripgrep

AndrewGaspar committed:
Merge remote-tracking branch 'upstream/develop' into cargo-package

Commit:
Add CI test for building ripgrep

AndrewGaspar commented:
Added a test in CI for building ripgrep - Rust is installed using a GitHub Action.

@adamjstewart I think this is ready for review.

Commit:
Fix regression introduced in merge for microarchitecture.py

AndrewGaspar commented on var/spack/repos/builtin/packages/rust/package.py:
```diff
@@ -541,13 +523,35 @@ def configure(self, spec, prefix):
 
-    def build(self, spec, prefix):
-        python('./x.py', 'build', extra_env={
+    def install(self, spec, prefix):
```
This change includes a couple small changes to the Rust package:
1) Starts building Rust with optimizations for the target microarchitecture
2) Removes the separate build phase and collapses it all into install.

AndrewGaspar commented on var/spack/repos/builtin/packages/rust/package.py:
```diff
@@ -541,13 +523,35 @@ def configure(self, spec, prefix):
+    def setup_dependent_package(self, module, dependent_spec):
+        """Called before cargo modules' build() methods."""
+
+        module.cargo = Executable(join_path(self.prefix.bin, 'cargo'))
```
This makes it so dependent packages that are using cargo directly will supply the `--locked` and `--offline` flags to help enforce the use of Spack's mechanisms for caching Cargo dependencies.

AndrewGaspar commented on lib/spack/llnl/util/cpu/microarchitectures.json:
```diff
@@ -1093,7 +1096,7 @@
           "versions": "3.9:",
           "family": "ppc64le",
-          "name": "power8",
+          "name": "pwr8",
```
As far as I can tell, clang treats "power8" and "pwr8" as synonyms, but LLVM (and, transitively, Rust) only recognizes "pwr8", so this change prefers the abbreviated version. Somebody with a better understanding of LLVM and microarchitectures may want to take a look, though.

AndrewGaspar commented on lib/spack/llnl/util/cpu/microarchitectures.json:
```diff
@@ -1012,6 +1012,7 @@
         },
         "clang": {
           "versions": "3.9:",
+          "name": "pwr7",
```
As far as I can tell, all of these microarchitectures are just as applicable to the non-LE PPC64 architecture, so I've supplied them here.

AndrewGaspar commented on lib/spack/llnl/util/cpu/microarchitecture.py:
```diff
@@ -191,23 +191,12 @@ def to_dict(self, return_list_of_items=False):
-        """
-        # If we don't have information on compiler at all
-        # return an empty string
+    def compiler_entry(self, compiler, version):
```
`compiler_entry` is broken out so that Rust packages can extract the microarchitecture names for "clang" since the arch names in clang are applicable to Rust, too.

AndrewGaspar commented on lib/spack/llnl/util/cpu/microarchitectures.json:
```diff
@@ -1174,7 +1177,8 @@
           {
             "versions": "5:",
-            "flags": "-mcpu=thunderx2t99"
+            "name": "thunderx2t99",
```
I decided to just re-use the clang arch names for Rust rather than try to define a complete separate list of flags. So therefore, the names must be exposed via the "name" key. Let me know how you feel about this approach.

AndrewGaspar commented on lib/spack/spack/stage.py:
```diff
@@ -735,6 +739,188 @@ def cache_local(self):
         tty.msg("Sources for DIY stages are not cached")
 
 
+class CargoStage(object):
```
This doesn't just get applied for those packages that derive `CargoPackage`, but any package that uses the `cargo_manifest` directive. This is to support caching of dependencies for packages that use Cargo internally, but where Cargo is not the top-level package.

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

AndrewGaspar commented:
ping @adamjstewart

adamjstewart commented:
No internet access at the moment, I'll try to review this when my internet is working again.

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

Commit:
Add Rust versions 1.45.1 and 1.44.1

Commit:
Merge branch 'rust-1.45.1' into cargo-package

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

AndrewGaspar committed:
Add cbindgen

AndrewGaspar committed:
Merge remote-tracking branch 'upstream/develop' into cargo-package

adamjstewart commented on var/spack/repos/builtin/packages/rust/package.py:
```diff
@@ -569,13 +551,35 @@ def configure(self, spec, prefix):
-    def build(self, spec, prefix):
-        python('./x.py', 'build', extra_env={
+    def install(self, spec, prefix):
+        jobs = spack.config.get('config:build_jobs') if self.parallel else 1
```
I think you want to use `make_jobs` here, it's essentially this.

adamjstewart commented on var/spack/repos/builtin/packages/jql/package.py:
```diff
@@ -0,0 +1,27 @@
+    )
+
+    version('master', branch='master')
+    version('2.6.1', sha256='849810db82ae770c9e9c64ee5a739ef624b67cf5e37ae806ce62fccee0cb8220')
```
I usually put versions before variants since the versions are more frequently updated.

adamjstewart commented on var/spack/repos/builtin/packages/git-delta/package.py:
```diff
@@ -0,0 +1,31 @@
+    def setup_build_environment(self, env):
+        # onig-sys environment
+        env.append_flags('RUSTONIG_DYNAMIC_LIBONIG', '1')
+        if '^llvm' in self.spec:
```
Doesn't this apply to all versions?

adamjstewart commented on lib/spack/spack/build_systems/cargo.py:
```diff
@@ -0,0 +1,247 @@
+            "LTO."
+    )
+
+    depends_on('rust', type='build', when='~prefer_dynamic')
+    depends_on('rust', type=('build', 'link'), when='+prefer_dynamic')
```
I'm not actually sure if this works, I've seen problems with changing the deptype based on variants.

scheibelp replied:
@adamjstewart you are generally correct but this does work when the constraints are mutually exclusive. An example of something that would not work:

```
depends_on(rust, type=build)
depends_on(rust, type=(build, link), when=+foo)
```

AndrewGaspar replied:
I added a comment describing why this is ok.

adamjstewart commented on lib/spack/spack/build_systems/cargo.py:
```diff
@@ -0,0 +1,247 @@
+
+    @property
+    def cargo(self):
+        cargo = Executable(join_path(self.spec['rust'].prefix.bin, 'cargo'))
```
```suggestion
        cargo = Executable(self.spec['rust'].prefix.bin.cargo)
```

adamjstewart commented on lib/spack/spack/build_systems/cargo.py:
```diff
@@ -0,0 +1,247 @@
+
+    @property
+    def rustc(self):
+        return Executable(join_path(self.spec['rust'].prefix.bin, 'rustc'))
```
```suggestion
        return Executable(self.spec['rust'].prefix.bin.rustc)
```

adamjstewart commented on lib/spack/spack/build_systems/cargo.py:
```diff
@@ -0,0 +1,247 @@
+
+    def build(self, spec, prefix):
+        """cargo build"""
+        jobs = spack.config.get('config:build_jobs') if self.parallel else 1
```
Use `make_jobs` here, you'll probably need `inspect` to access it. See other build systems for how to do this.

adamjstewart commented on lib/spack/spack/build_systems/cargo.py:
```diff
@@ -0,0 +1,247 @@
+
+        # This is not implemented using "cargo install" because we want to
+        # install libraries, too.
+        so_ext = 'dylib' if 'platform=darwin' in self.spec else 'so'
```
This should already be defined elsewhere in Spack

AndrewGaspar replied:
So there's [dso_suffix](https://github.com/spack/spack/blob/develop/lib/spack/spack/build_environment.py#L96). However, I prefer my version because it's derived from the spec's platform, not from the host's platform. It probably doesn't matter since spack doesn't really currently support cross compiling between platforms all that well, but maybe someday?

adamjstewart commented on lib/spack/spack/build_systems/cargo.py:
```diff
@@ -0,0 +1,247 @@
+        so_ext = 'dylib' if 'platform=darwin' in self.spec else 'so'
+
+        # This loop attempts to install all
+        for t in self.workspace_targets():
```
Let's use a more informative variable name like `target`

adamjstewart reviewed (commented):
Sorry I haven't gotten a chance to review this until now.

Most of the package changes, build system docs, and build system stuff looks good to me. The changes to fetching and archspec stuff will need to reviewed by other maintainers. Maybe @alalazo @becker33 @scheibelp @tgamblin would want to review this.

AndrewGaspar committed:
Accept err_msg to CargoStage.fetch

AndrewGaspar committed:
new package versions

Commit:
Add Rust 1.46.0

Commit:
Merge remote-tracking branch 'agaspar/spack-rust-1.46.0' into cargo-package

AndrewGaspar committed:
Fix bug in default extra_targets

hhirtz commented:
Hi Andrew!

Thank you very much for this PR.  I tested the installation of ripgrep without internet (from a local spack mirror), and it works!  I'll test it with a library that exports a C interface and come back with the results (there isn't one in spack's builtin repo already is there?)

Rust libraries (with a rust interface) can be made available in a local cargo registry [0], so that cargo manifests and build commands don't have to be changed to work without internet.  Then spack shouldn't be needed for this use-case.

[0] https://github.com/ChrisGreenaway/cargo-local-registry
https://doc.rust-lang.org/cargo/reference/source-replacement.html

AndrewGaspar commented:
> Then spack shouldn't be needed for this use-case.

Of course - but the goal of this PR is to make sure cargo packages can be integrated easily within a larger toolchain that can be managed completely by spack.

cedricchevalier19 commented:
I really hope to see this PR merged soon. It would be awesome to use our favorite rust tools easily on offline clusters !

AndrewGaspar committed:
Merge remote-tracking branch 'upstream/develop' into cargo-package

AndrewGaspar committed:
Use make_jobs

AndrewGaspar committed:
t -> target

AndrewGaspar committed:
Move version directives towards top of Rust packages

AndrewGaspar committed:
Remove check for llvm dependency from git-delta

AndrewGaspar committed:
Add comment about multiple depends_on('rust') in cargo.py

AndrewGaspar committed:
use . syntax for path construction

AndrewGaspar commented:
Working on adding unit tests to increase code coverage... I didn't realize that `linux_build_tests.yaml` doesn't contribute to code coverage.

AndrewGaspar commented:
I'm having a hard time writing a unit test. Is there any way to specify a configuration for a test? I want my test to run as-if it's concretizing with a recent version of clang so that I can test the target-cpu code.

adamjstewart commented:
We don't have any coverage for any other build system in Spack, so I wouldn't require coverage for your build system in order to merge this PR. At some point I would like to add coverage for our existing build systems, but I haven't had a chance to yet. @alalazo is our testing wizard, so he may have more comments about the best way to do this. We used to have coverage for linux build tests, but as far as I know, uploading coverage for several tests with codecov is currently broken.

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

AndrewGaspar committed:
Merge remote-tracking branch 'upstream/develop' into cargo-package

AndrewGaspar committed:
Update starship

AndrewGaspar committed:
Merge branch 'cargo-package' of https://github.com/AndrewGaspar/spack into cargo-package

hhirtz commented:
The timing for dependency vendoring does not allow patches on the cargo
manifest:  `cargo vendor` is called in the stage phase before any patch
can be applied.

I looked into it for a bit and there doesn't seem to be a trivial fix
for this (very new to spack, so i could be wrong :).  In this case this
should be documented with `CargoPackage`, otherwise it will lead to
weird bugs due to cargo using the unpatched manifest.

Example:

```toml
[target.'cfg(windows)'.dependencies]
unused-arch-specific-dep = { path = "../non-existent-dir" }
```

Edit: actually other files can still be patched fine.

Commit:
Merge remote-tracking branch 'origin/develop' into cargo-package

tgamblin committed:
update CHANGELOG.md for v0.16.0

tgamblin committed:
bump version number to 0.16.0

Commit:
Merge tag 'v0.16.0' into cargo-package

AndrewGaspar committed:
starship 0.48.0

hhoeflin commented:
Any news on this?

AndrewGaspar commented:
I'm no longer working in HPC, so I'm unlikely to do the work to get this merged. Somebody else will need to take it over.

alalazo commented:
Closing, since there has been no activity in a while (and there are a lot of conflicts). Feel free to reopen if you want to continue work on this.
